pkg/common-controller/groupsnapshot_controller.go

```diff
@@ -9,3 +22,236 @@ func (ctrl *csiSnapshotCommonController) storeGroupSnapshotUpdate(groupsnapshot
+		return nil, groupSnapshot, err
+	}
+
+	pvDriver, err := ctrl.pvDriverFromGroupSnapshot(groupSnapshot)
+	if err != nil {
+		klog.Errorf("failed to get pv csi driver from group snapshot %s/%s: %q", groupSnapshot.Namespace, groupSnapshot.Name, err)
+		return nil, groupSnapshot, err
+	}
+
+	defaultClasses := []*crdv1alpha1.VolumeGroupSnapshotClass{}
+	for _, class := range list {
+		if utils.IsDefaultAnnotation(class.ObjectMeta) && pvDriver == class.Driver {
+			defaultClasses = append(defaultClasses, class)
+			klog.V(5).Infof("get defaultGroupClass added: %s, driver: %s", class.Name, pvDriver)
+		}
+	}
+	if len(defaultClasses) == 0 {
+		return nil, groupSnapshot, fmt.Errorf("cannot find default group snapshot class")
+	}
+	if len(defaultClasses) > 1 {
+		klog.V(4).Infof("get DefaultClass %d defaults found", len(defaultClasses))
+		return nil, groupSnapshot, fmt.Errorf("%d default snapshot classes were found", len(defaultClasses))
+	}
+	klog.V(5).Infof("setDefaultSnapshotClass [%s]: default VolumeSnapshotClassName [%s]", groupSnapshot.Name, defaultClasses[0].Name)
+	groupSnapshotClone := groupSnapshot.DeepCopy()
+	groupSnapshotClone.Spec.VolumeGroupSnapshotClassName = &(defaultClasses[0].Name)
+	newGroupSnapshot, err := ctrl.clientset.GroupsnapshotV1alpha1().VolumeGroupSnapshots(groupSnapshotClone.Namespace).Update(context.TODO(), groupSnapshotClone, metav1.UpdateOptions{})
+	if err != nil {
+		klog.V(4).Infof("updating VolumeSnapshot[%s] default class failed %v", utils.GroupSnapshotKey(groupSnapshot), err)
+	}
+	_, updateErr := ctrl.storeGroupSnapshotUpdate(newGroupSnapshot)
+	if updateErr != nil {
+		// We will get an "snapshot update" event soon, this is not a big error
+		klog.V(4).Infof("setDefaultSnapshotClass [%s]: cannot update internal cache: %v", utils.GroupSnapshotKey(groupSnapshot), updateErr)
+	}
+
+	return defaultClasses[0], newGroupSnapshot, nil
+}
+
+// pvDriverFromGroupSnapshot is a helper function to get the CSI driver name from the targeted PersistentVolume.
+// It looks up the PVC from which the snapshot is specified to be created from, and looks for the PVC's corresponding
+// PV. Bi-directional binding will be verified between PVC and PV before the PV's CSI driver is returned.
+// For an non-CSI volume, it returns an error immediately as it's not supported.
+func (ctrl *csiSnapshotCommonController) pvDriverFromGroupSnapshot(groupSnapshot *crdv1alpha1.VolumeGroupSnapshot) (string, error) {
+	pv, err := ctrl.getVolumeFromVolumeGroupSnapshot(groupSnapshot)
+	if err != nil {
+		return "", err
+	}
+	// supports ONLY CSI volumes
+	if pv.Spec.PersistentVolumeSource.CSI == nil {
+		return "", fmt.Errorf("snapshotting non-CSI volumes is not supported, snapshot:%s/%s", groupSnapshot.Namespace, groupSnapshot.Name)
+	}
+	return pv.Spec.PersistentVolumeSource.CSI.Driver, nil
+}
+
+func (ctrl *csiSnapshotCommonController) getVolumeFromVolumeGroupSnapshot(groupSnapshot *crdv1alpha1.VolumeGroupSnapshot) (*v1.PersistentVolume, error) {
```

Comment:
I don't think we need this.  VolumeGroupSnapshot and PV are not one-on-one relationship.  We can probably use the existing function from snapshot controller.

Reply:
Yes you are right. I actually updated these functions in another commit that is not a part of this PR. I've pushed that here now - basically returning a list of volumes from the group snapshot. 

There's actually a lot of common functions that can potentially be refactored, but as we discussed offline I kept it separate to not touch the VolumeSnapshot code. Do you think its better to do some refactoring?

Reply:
I thought you might be able to call an existing function directly without any modifications.  However if you need to refactor in order to call an existing function, then just create a new one.

Reply:
There's a certain degree of refactoring needed. Ill leave this as is for now, we can look into the refactoring later